### PR TITLE
feat: Add delete and finalizer to RAGEngine

### PR DESCRIPTION
**Pull request by bangqipropel**
**Reason for Change**:
Add delete and finalizer to RAGEngine

**Requirements**

- [ ] added unit tests and e2e tests (if applicable).

**Issue Fixed**:

**Notes for Reviewers**:

### CODE REVIEW

**Review by ishaansehgal99 — APPROVED** [review]
LGTM

File: pkg/controllers/ragengine_gc_finalizer.go [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+	staleWObj := ragEngineObj.DeepCopy()
+	staleWObj.SetFinalizers(nil)
+	if updateErr := c.Update(ctx, staleWObj, &client.UpdateOptions{}); updateErr != nil {
+		klog.ErrorS(updateErr, "failed to remove the finalizer from the ragengine",
+			"ragengine", klog.KObj(ragEngineObj), "ragengine", klog.KObj(staleWObj))
+		return ctrl.Result{}, updateErr
+	}
+	klog.InfoS("successfully removed the ragengine finalizers",
+		"ragengine", klog.KObj(ragEngineObj))
+	controllerutil.RemoveFinalizer(ragEngineObj, consts.RAGEngineFinalizer)
```
Comment (zhuangqh) [inline pkg/controllers/ragengine_gc_finalizer.go:62]: why remove the finalizer again here?

File: pkg/controllers/ragengine_gc_finalizer.go [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+		for i := range ncList.Items {
+			if deleteErr := c.Delete(ctx, &ncList.Items[i], &client.DeleteOptions{}); deleteErr != nil {
+				klog.ErrorS(deleteErr, "failed to delete the nodeClaim", "nodeClaim", klog.KObj(&ncList.Items[i]))
+				return ctrl.Result{}, deleteErr
+			}
+		}
+	}
+
+	staleWObj := ragEngineObj.DeepCopy()
+	staleWObj.SetFinalizers(nil)
```
Comment (zhuangqh) [inline pkg/controllers/ragengine_gc_finalizer.go:54]: we should only remove the finalizer which is set by this controller, or we may remove the finalizer set by user by accident.

File: pkg/controllers/ragengine_gc_finalizer.go (hunk 1 above)
Comment (bangqipropel) [reply pkg/controllers/ragengine_gc_finalizer.go:62]: Why we say "again" here?

File: pkg/controllers/ragengine_gc_finalizer.go (hunk 2 above)
Comment (bangqipropel) [reply pkg/controllers/ragengine_gc_finalizer.go:54]: I think from 53-62 all about removing the finalizer?

Comment (zhuangqh) [reply pkg/controllers/ragengine_gc_finalizer.go:54]: Suppose finalizers is ["kaito/a", "from-user/b"], just remove the finalizer set by this controller (kaito/b), and set finalizers to ["from-user/b"].
you can take a look at this code https://github.com/openkruise/kruise/blob/master/pkg/controller/ephemeraljob/ephemeraljob_utils.go#L256

File: pkg/controllers/ragengine_gc_finalizer.go (hunk 1 above)
Comment (zhuangqh) [reply pkg/controllers/ragengine_gc_finalizer.go:62]: you have made an update to the apisever using `staleWObj`.
and update to `ragEngineObj` again

**Review by zhuangqh — APPROVED** [review]

**Review by Fei-Guo — APPROVED** [review]

**Pull request merged**